doc/internal/man3/ossl_paramtmpl_init.pod

```diff
@@ -0,0 +1,180 @@
+                                   void *buf, size_t bsize);
+
+
+=head1 DESCRIPTION
+
+A collection of utility functions that simplify the creation of OSSL_PARAM
+arrays.  The B<TYPE> names are as per L<OSSL_PARAM_int(3)>.
+
+ossl_paramtmpl_init() initialises the OSSL_PARAMTMPL structure so that values
+can be added.
+Any existing values are cleared.
+
+ossl_paramtmpl_to_param() converts a built up OSSL_PARAMTMPL structure
+B<tmpl> into an allocated OSSL_PARAM array.
+The OSSL_PARAM array and all associated storage can be freed by calling
+OPENSSL_free() or, if it was private, OPENSSL_secure_free().
+
+ossl_paramtmpl_to_param_ex() behaves like ossl_paramtmpl_to_param(), except that
+no additional memory is allocated.
+An OSSL_PARAM array of at least B<param_n> elements is passed in as B<params>.
+The auxilary storage for the parameters is a block of memory pointed to
+by B<data> of at least B<data_n> bytes in size.
+
+ossl_paramtmpl_push_TYPE() are a series of functions which will create
+OSSL_PARAM objects of the specified size and correct type for the B<val>
+argument.
+B<val> is stored by value and an expression or auto variable can be used.
+
+ossl_paramtmpl_push_BN() is a function that will create an OSSL_PARAM object
+that holds the specified BIGNUM B<bn>.
+If B<bn> is marked as being securely allocated, the
+ossl_paramtmpl_set_private() function is called which allocates the
+resulting OSSL_PARAM array and associated storage securely.
+The B<bn> argument is stored by reference and the underlying BIGNUM object
+must exist until after ossl_paramtmpl_to_param() has been called.
+
+ossl_paramtmpl_push_utf8_string() is a function that will create an OSSL_PARAM
+object that references the UTF8 string specified by B<buf>.
+If the length of the string, B<bsize>, is zero then it will be calculated.
+The string that B<buf> points to is stored by reference and must remain in
+scope until after ossl_paramtmpl_to_param() has been called.
+
+ossl_paramtmpl_push_octet_string() is a function that will create an OSSL_PARAM
+object that references the octet string specified by B<buf> and <bsize>.
+The memory that B<buf> points to is stored by reference and must remain in
+scope until after ossl_paramtmpl_to_param() has been called.
+
+ossl_paramtmpl_push_utf8_ptr() is a function that will create an OSSL_PARAM
+object that references the UTF8 string specified by B<buf>.
+If the length of the string, B<bsize>, is zero then it will be calculated.
+The string B<buf> points to is stored by reference and must remain in
+scope until the OSSL_PARAM array is freed.
+
+ossl_paramtmpl_push_octet_ptr() is a function that will create an OSSL_PARAM
+object that references the octet string specified by B<buf>.
+The memory B<buf> points to is stored by reference and must remain in
+scope until the OSSL_PARAM array is freed.
+
+=head1 RETURN VALUES
+
+ossl_paramtmpl_to_param() return the allocated OSSL_PARAM array, or NULL on
+error.
```

Comment:
Should it raise an error on the error stack?

Reply:
It could, malloc failed is the only possible error currently.

For ossl_paramtmpl_to_param_ex the error is from the bounds checking.

Reply:
Done